backend/endpoints_rooms.go

```diff
@@ -1,512 +1,533 @@
+			write <- RoomInfoMessageOutgoing{
+				Type: "room_info",
+				Data: RoomInfoMessageOutgoingData{
+					ID:         id,
+					CreatedAt:  &createdAt,
+					ModifiedAt: &modifiedAt,
+					Type:       body.Type,
+					Target:     body.Target,
+				},
+			}
+			return true
+		})
+	}
+	w.Write([]byte("{\"success\":true}"))
+}
+
+func GetRoomSubtitleEndpoint(w http.ResponseWriter, r *http.Request) {
+	if user, _ := IsAuthenticatedHTTP(w, r); user == nil {
+		return
+	} else if r.URL.Query().Get("name") == "" {
+		http.Error(w, errorJson("Name cannot be empty!"), http.StatusBadRequest)
+		return
+	}
+
+	var subtitle string
+	err := findSubtitleStmt.QueryRow(r.PathValue("id"), r.URL.Query().Get("name")).Scan(&subtitle)
+	if errors.Is(err, sql.ErrNoRows) {
+		http.Error(w, errorJson("Subtitles or room not found!"), http.StatusNotFound)
+		return
+	} else if err != nil {
+		handleInternalServerError(w, err)
+		return
+	}
+	w.Write([]byte(subtitle))
+}
+
+func CreateRoomSubtitleEndpoint(w http.ResponseWriter, r *http.Request) {
+	if user, _ := IsAuthenticatedHTTP(w, r); user == nil {
+		return
+	} else if r.URL.Query().Get("name") == "" {
+		http.Error(w, errorJson("Name cannot be empty!"), http.StatusBadRequest)
+		return
+	}
+
+	body, err := io.ReadAll(http.MaxBytesReader(w, r.Body, 1024*1024)) // 1 MB limit
+	if err != nil || len(body) == 0 {
+		http.Error(w, errorJson("Unable to read body!"), http.StatusBadRequest)
+		return
+	} else if len(body) == 1024*1024 {
+		http.Error(w, errorJson("Body too large!"), http.StatusRequestEntityTooLarge)
+		return
+	}
+
+	result, err := insertSubtitleStmt.Exec(r.PathValue("id"), r.URL.Query().Get("name"), body)
+	if pqErr, ok := err.(*pq.Error); ok && pqErr.Code.Class() == "23503" {
+		http.Error(w, errorJson("Room does not exist!"), http.StatusConflict)
+		return
+	} else if err != nil {
+		handleInternalServerError(w, err)
+		return
+	} else if rows, err := result.RowsAffected(); err != nil || rows != 1 {
+		handleInternalServerError(w, err)
+		return
+	}
+
+	// Send message to all room members about the change
+	if members, ok := roomMembers.Load(r.PathValue("id")); ok {
+		members.Range(func(write chan<- interface{}, userId uuid.UUID) bool {
+			write <- SubtitleMessageOutgoing{Type: "subtitle", Data: []string{r.URL.Query().Get("name")}}
+			return true
+		})
+	}
+
+	w.Write([]byte("{\"success\":true}"))
+}
+
+type AuthMessageIncoming struct {
+	Token     string `json:"token"`
+	Reconnect bool   `json:"reconnect"` // If this is a reconnect
+}
+
+type GenericMessage struct {
+	Type string `json:"type"`
+}
+
+type ChatMessageIncoming struct {
+	Type string `json:"type"` // chat
+	Data string `json:"data"`
+}
+
+type PingPongMessageBi struct {
+	Type      string `json:"type"` // ping if incoming, pong if outgoing
+	Timestamp int    `json:"timestamp"`
+}
+
+type TypingIndicatorMessageBi struct {
+	Type string `json:"type"` 
+	Data map[string]int64 `json:"data"`

```

Comment:
This could just be `Timestamp int` and `TypingIndicatorMessageIncoming`

The Outgoing equivalent should have the `username` of the user typing set by the backend